### PR TITLE
[UL&S] Login with site address: Jetpack not installed or not connected

### LINKED ISSUE #3130: [UL&S] UI. Updates to error states: jetpack not installed
Update Error screens:

Screenshot 2020-11-11 at 11 12 59 AM

### LINKED ISSUE #3131: [UL&S] UI. Jetpack required
Screenshot 2020-11-11 at 11 14 11 AM

### PR DESCRIPTION
**Pull request by ctarda**
Closes #3130 
Closes #3131 

Ref: p91TBi-3ON-p2

⚠️ Please note this PR is not against develop, but against the feature branch for UL&S 🙇

##Design

## Changes

In action:

Screenshots:

| | Light Appearance | Dark Appearance |
| ---- | ---- | ---- |
| iPhone SE |  |  |
| iPhone 12 |  |  |
| iPhone 12 Pro Max |  |  |

⚠️  Adding the help button requires another update to WPAuthenticator, so I have deferred it to a separate issue: #3214

## How it works
The heavy lifting is done by WordPressAuthenticator. What we do is provide a View Controller to the library, and let it inset it in whatever way it makes sense. See p91TBi-3ON-p2 for more context.

This PR:
* Points the Podfile to a specific beta release of WPAuthenticator.
* Adds a new view controller (ULErrorViewController). All error messages for UL&S seem to be modeled using the same UI pattern, so I decided to go "generic", and attempt to implement all these special screens (that contain an image, text, a text button and two larger buttons) using the same view controller
* The way to customize this ULErrorViewController is by injecting a view model that provides the content (i.e. text, button titles) and that handles button actions.
* Adds unit tests

## Testing
* Checkout the branch. Run `bundle exec pod install`
* Log out if necessary
* Attempt to Log In with Store Address. Enter a store address of a site that does not have Jetpack installed or does not have Jetpack connected.
* Notice if the error screen in shown. 
* The "What is Jetpack" button should load a web view pointing to this article: https://jetpack.com/about/ * The "See instructions" button should load a web view pointing to: https://docs.woocommerce.com/document/jetpack-setup-instructions-for-the-woocommerce-mobile-app/
* The "Refresh after install" button should navigate to the screen where users enter the site address. It is not actually possible to refresh this screen, so there isn't much more we can do, as far as I can tell. Better solutions are welcome!

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/WooCommerceTests/Extensions/StringWooTests.swift [hunk 1]
```diff
@@ -21,4 +21,25 @@ final class StringWooTests: XCTestCase {
...
+
+        XCTAssertEqual(url.trimHTTPScheme(), expectedResult)
+    }
+
+    func testStringIsNotTrimmerIfItIsNotAURL() {
+        let expectedResult = "regular string"
+        let notAURL = expectedResult
+
+        XCTAssertEqual(notAURL.trimHTTPScheme(), expectedResult)
+    }
```
Comment (pmusolino) [inline WooCommerce/WooCommerceTests/Extensions/StringWooTests.swift]: Nice that you added the tests for `trimHTTPScheme()` 💯 

Nit: can you rename the tests following our [new convention](https://github.com/woocommerce/woocommerce-ios/blob/develop/docs/naming-conventions.md#test-methods)?

File: WooCommerce/WooCommerceTests/Extensions/IconsTests.swift [hunk 2]
```diff
@@ -168,6 +168,10 @@ final class IconsTests: XCTestCase {
         XCTAssertNotNil(UIImage.loginSiteAddressInfoImage)
     }
 
+    func testLoginJetpackErrorIconIsNotNil() {
```
Comment (pmusolino) [inline WooCommerce/WooCommerceTests/Extensions/IconsTests.swift]: Nit: rename the test following our [new convention](https://github.com/woocommerce/woocommerce-ios/blob/develop/docs/naming-conventions.md#test-methods)

File: WooCommerce/Classes/Authentication/Navigation Exceptions/ULErrorViewController.swift [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import SafariServices
+
+final class ULErrorViewController: UIViewController {
+    private let viewModel: ULErrorViewModel
+
+    @IBOutlet private var primaryButton: NUXButton!
+    @IBOutlet private var secondaryButton: NUXButton!
+    @IBOutlet private var imageView: UIImageView!
+    @IBOutlet weak var errorMessage: UILabel!
+    @IBOutlet weak var extraInfoButton: UIButton!
```
Comment (pmusolino) [inline WooCommerce/Classes/Authentication/Navigation Exceptions/ULErrorViewController.swift]: nit: `private`

File: WooCommerce/WooCommerceTests/Authentication/ULErrorViewControllerTests.swift [hunk 4]
```diff
@@ -0,0 +1,158 @@
+import XCTest
+@testable import WooCommerce
+
+final class ULErrorViewControllerTests: XCTestCase {
```
Comment (pmusolino) [inline WooCommerce/WooCommerceTests/Authentication/ULErrorViewControllerTests.swift:4]: 💯

**Review by pmusolino — APPROVED** [review]
It looks good to me :shipit: I left just some nits.

About this point:
> The "Refresh after install" button should navigate to the screen where users enter the site address. It is not actually possible to refresh this screen, so there isn't much more we can do, as far as I can tell. Better solutions are welcome!

Do you think that would be possible in some way to do a `popViewController`, and retry on the same URL? I think it would be better than endlessly presenting the same views.

**Review by Garance91540 — COMMENTED** [review]
Hi Cesar, great progress!

A few notes I found:
- What is Jetpack opens the "[what is Jetpack modal](https://d.pr/i/wWMtu2)". Selecting "Learn more about Jetpack" on that modal opens up the [about Jetpack page](https://jetpack.com/about/)
- The "Refresh after install" CTA was meant for users who've installed Jetpack after reading the error message to be able to refresh the screen, and jump directly to the email step. In Android, we ended up changing this "Refresh after install" CTA with "Log in with another account" which takes you back to the app landing screen. Maybe it's easier to go this way, so I updated designs accordingly

File: WooCommerce/WooCommerceTests/Extensions/StringWooTests.swift (hunk 1 above)
Comment (ctarda) [reply WooCommerce/WooCommerceTests/Extensions/StringWooTests.swift]: Right! Thanks for pointing that out. I considered it, but decided to follow the convention that the rest of this particular XCTestCase subclass was following. 

I'll update!

**Comment by ctarda** [comment]
Thanks for the review @Garance91540 @pmusolino 

This is my first real large PR after a year away from the project, and there are a lot of moving parts (mostly WPAuthenticator) so as much as I'd like to make the review easy for you I am afraid that I'll miss things here and there. That's all to say that I appreciate the thorough reviews 🙇‍♂️

Anyway, I've pushed an update that incorporates Paolo's feedback about the code. There is also some new code (sorry Paolo):
* A new extension on FancyAlertViewController to encapsulate the alert. This could probably be generalized a bit as we go implementing more screens in the new login flow. Let's see how it goes.
* An update to the JetpackErrorViewModel to change the copy and logic of the secondary action button (the one at the very bottom) as well as the logic of the auxiliary button (the one what is Jetpack button)
* The pod file points to a new brach in WPAuthenticator (so that we can navigate to the root of the login flow) There is a PR in WPAuthenticator for that: wordpress-mobile/WordPressAuthenticator-iOS#524

This is how the interactions with this screen look now:



And a couple of screenshots in light and dark appearance:

| Light mode | Dark mode |
| ---- | ---- |
|  |  |
|  |  |

**Review by Garance91540 — APPROVED** [review]
Looking great! 👏👏👏

File: WooCommerce/Classes/Authentication/Navigation Exceptions/ULErrorViewController.swift [hunk 5]
```diff
@@ -0,0 +1,122 @@
+
```
Comment (pmusolino) [inline WooCommerce/Classes/Authentication/Navigation Exceptions/ULErrorViewController.swift]: nit: remove this extra space

**Review by pmusolino — APPROVED** [review]
LGTM :shipit:

**Comment by ctarda** [comment]
Thanks folks. Waiting for CI to pass. I'll merge as soon as we get the ✅

**Pull request merged**